### ISSUE TITLE
Waypoints

### ISSUE THREAD

**Issue by cKnei** [SkyblockerMod/Skyblocker#968]
### What happened?

Okay this was kind of on me, but at the same time I thought this might be something to address

When [loading Waypoints](https://github.com/SkyblockerMod/Skyblocker/blob/master/src/main/java/de/hysky/skyblocker/skyblock/waypoint/Waypoints.java#L60), if it errors, the next time the game is loaded, the entire file is cleared of all waypoints due to overwriting it with the empty Waypoint List that is saved in memory

### Minecraft Version

1.21.0/1.21.1

### Skyblocker Version

skyblocker-1.20.6+1.21.2

**Comment by kevinthegreat1** [comment]
Simple fix by applying the following diff

```diff
    public static void loadWaypoints() {
-        waypoints.clear();
        try (BufferedReader reader = Files.newBufferedReader(waypointsFile)) {
            List<WaypointCategory> waypoints = CODEC.parse(JsonOps.INSTANCE, SkyblockerMod.GSON.fromJson(reader, JsonArray.class)).resultOrPartial(LOGGER::error).orElseThrow();
+            Waypoints.waypoints.clear();
            waypoints.forEach(waypointCategory -> Waypoints.waypoints.put(waypointCategory.island(), waypointCategory));
        } catch (Exception e) {
            LOGGER.error("[Skyblocker Waypoints] Encountered exception while loading waypoints", e);
        }
    }
```

to 


https://github.com/SkyblockerMod/Skyblocker/blob/3b344cdeee980a3eaf8609b983a4aafb189a6f46/src/main/java/de/hysky/skyblocker/skyblock/waypoint/Waypoints.java#L62-L70

**Comment by kevinthegreat1** [comment]
I completely misunderstood your problem. Please ignore the above fix. The best solution for this is to probably backup the old waypoints file when saving a new one.

**Comment by cKnei** [comment]
> I completely misunderstood your problem. Please ignore the above fix. The best solution for this is to probably backup the old waypoints file when saving a new one.

IT was also on me, I was not exactly paying attention when applying the diff, oops.
But I would like to ask, would it make sense to not save to a file if there are no waypoints at all?

eg.

```diff
public static void saveWaypoints(MinecraftClient client) {
+    if (waypoints.isEmpty()) return;
    try (BufferedWriter writer = Files.newBufferedWriter(waypointsFile)) {
        JsonElement waypointsJson = CODEC.encodeStart(JsonOps.INSTANCE, List.copyOf(waypoints.values())).resultOrPartial(LOGGER::error).orElseThrow();
        SkyblockerMod.GSON.toJson(waypointsJson, writer);
        LOGGER.info("[Skyblocker Waypoints] Saved waypoints");
    } catch (Exception e) {
        LOGGER.error("[Skyblocker Waypoints] Encountered exception while saving waypoints", e);
    }
}
```

**Comment by kevinthegreat1** [comment]
> It was also on me, I was not exactly paying attention when applying the diff, oops. But I would like to ask, would it make sense to not save to a file if there are no waypoints at all?

Then you can't delete all the waypoints.

**Comment by cKnei** [comment]
> > It was also on me, I was not exactly paying attention when applying the diff, oops. But I would like to ask, would it make sense to not save to a file if there are no waypoints at all?
> 
> Then you can't delete all the waypoints.

Well based on what I understand of the Waypoint class, it would only delete it from the the MultiMap, which should not effect the actual waypoint file in any way. 

I probably am just be overthinking this and, yeah honestly keeping a copy is probably the best idea.

**Comment by kevinthegreat1** [comment]
> Well based on what I understand of the Waypoint class, it would only delete it from the the MultiMap, which should not effect the actual waypoint file in any way.

If someone delets all the waypoints, under your proposal, then the file won't get saved. On the next launch, the old waypoints will be loaded again, even though the user has deleted all of them.

**Comment by cKnei** [comment]
> If someone delets all the waypoints, under your proposal, then the file won't get saved. On the next launch, the old waypoints will be loaded again, even though the user has deleted all of them.

You're absolutely right. Sorry didn't see that far ahead, but based off this it is all the solutions I had in mind.

In the end I will just stick to keeping a copy of my stuff, if that is all, feel free to close this issue
